nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -117,28 +128,19 @@ class BtActionNode : public BT::CoroActionNode
       throw std::runtime_error("Goal was rejected by the action server");
     }
 
-    auto future_result = action_client_->async_get_result(goal_handle_);
-    rclcpp::executor::FutureReturnCode rc;
-    do {
-      rc = rclcpp::spin_until_future_complete(node_, future_result, server_timeout_);
-      if (rc == rclcpp::executor::FutureReturnCode::TIMEOUT) {
-        on_server_timeout();
-
-        // We can handle a new goal if we're still executing
-        auto status = goal_handle_->get_status();
-        if (goal_updated_ && (status == action_msgs::msg::GoalStatus::STATUS_EXECUTING ||
-          status == action_msgs::msg::GoalStatus::STATUS_ACCEPTED ))
-        {
-          goal_updated_ = false;
-          goto new_goal_received;
-        }
-
-        // Yield to any other CoroActionNodes (coroutines)
-        setStatusRunningAndYield();
+    rclcpp::Rate r(20);
+    while (rclcpp::ok() && !goal_result_available) {
+      on_wait_for_result();
+      if (goal_updated_) {
+        goal_updated_ = false;
+        goto new_goal_received;
       }
-    } while (rc != rclcpp::executor::FutureReturnCode::SUCCESS);
 
-    result_ = future_result.get();
+      // Yield to any other nodes
+      setStatusRunningAndYield();
+      rclcpp::spin_some(node_);
```

Comment:
Spin some is not intuitive. It means, "you must spin at least once". If there is no work in the queue, it blocks until there is some work.

This would block the behavior tree.

Reply:
My understanding from this question I asked last year (https://answers.ros.org/question/296480/functional-difference-between-ros2-spin_some-spin_once-and-spin_until_future/) is that spin some will execute all in the queue, if the queue is empty it wont block

Reply:
I'd also be fine if somewhere else in the tree we were actually spinning this node that is created. No where is the node actually spun so I needed to add it somewhere that it was used. Other option is to have a thread created in the factory node that is spinning it constantly but this seems more controlled

Reply:
Yes, the new approach is to not use a non-spining node for the BT, but a spinning node. Then wait on futures instead of manually doing the spin_until_future_complete. Pending the verification of Carl's comment, I think this is OK until we switch to the new method.